File: printer.py
"""
Printer - a live view of the Klipper/Fluidd machine, under /api/printer.

The whole design here follows from one constraint: **Ops Deck is served over
HTTPS and the printer speaks plain HTTP on the LAN.** Browsers refuse to load
http:// subresources into an https:// page, so nothing on 10.0.0.x may be
referenced by the browser directly - an <img> or <iframe> pointing at the
printer renders an empty box on every device, with no error the user can see.

So there are exactly two ways across that line, and this module is one of them:

  - **The camera is proxied here.** /snapshot and /stream fetch from
    mjpg-streamer server-side and re-serve the bytes same-origin, which makes
    them ordinary HTTPS resources the browser is happy with.
  - **The Fluidd UI is fronted by its own `tailscale serve` listener** on
    :8444, because a single-page app cannot be usefully reverse-proxied under
    a path prefix - Fluidd requests /js/app.js absolutely, and rewriting an
    SPA's asset paths and its Moonraker websocket on the fly is a losing game.
    That URL is only reachable inside the tailnet.

Deliberately **not profile-scoped**: there is one physical printer and one
household, so this ignores X-Profile-Id the way /api/joint does. Nothing here
writes - no endpoint can move an axis, heat a nozzle or start a job. Read-only
by construction, so the worst a bug can do is show a stale temperature.

A powered-off printer is the normal case, not an error case. Every endpoint
degrades to a clean "offline" rather than a stack trace or a hung request.
"""
import json
import os
import threading
import urllib.error
import urllib.request

from flask import Blueprint, Response, jsonify

from api import require_token
from db import connect  # noqa: F401  - kept for symmetry with the other blueprints

printer = Blueprint("printer", __name__, url_prefix="/api/printer")

HOST = os.environ.get("OPSDECK_PRINTER_HOST", "10.0.0.131")
CAMERA_PORT = os.environ.get("OPSDECK_PRINTER_CAMERA_PORT", "8080")
# The HTTPS front for the Fluidd SPA. Empty disables the embedded dashboard
# and the UI falls back to its own status panel plus a plain link.
UI_URL = os.environ.get("OPSDECK_PRINTER_UI_URL",
                        "https://opsdeck.example.ts.net:8444")

MOONRAKER = f"http://{HOST}"
CAMERA = f"http://{HOST}:{CAMERA_PORT}"

# Short, because the printer being off must feel like "off", not like the tab
# hanging. A dead host on the same subnet refuses fast; this is the ceiling.
TIMEOUT = 5
SNAPSHOT_TIMEOUT = 8

# Each live stream holds a thread and an upstream socket for as long as the
# tab is open. One viewer is the real case; the cap stops a stack of forgotten
# phone tabs from quietly eating the app's worker threads.
MAX_STREAMS = 3
_stream_slots = threading.BoundedSemaphore(MAX_STREAMS)

# The Moonraker objects worth one round trip.
QUERY = ("print_stats&display_status&heater_bed&extruder"
         "&toolhead&virtual_sdcard&idle_timeout")


def _get_json(url, timeout=TIMEOUT):
    """GET and parse, or raise. Callers turn failure into 'offline'."""
    req = urllib.request.Request(url, headers={"User-Agent": "OpsDeck/1.0"})
    with urllib.request.urlopen(req, timeout=timeout) as res:
        return json.loads(res.read().decode())


def _num(value, fallback=0.0):
    try:
        return float(value)
    except (TypeError, ValueError):
        return fallback


def _heater(obj):
    return {
        "temp": round(_num(obj.get("temperature")), 1),
        "target": round(_num(obj.get("target")), 1),
        "power": round(_num(obj.get("power")), 2),
    }


def read_status():
    """
    One normalised snapshot of the machine, or an offline marker.

    Never raises: a printer that is unplugged, asleep, or mid-firmware-restart
    is an ordinary state for this tab to render.
    """
    try:
        info = _get_json(f"{MOONRAKER}/server/info")
    except Exception as e:
        return {"online": False,
                "error": f"{type(e).__name__}: {e}",
                "host": HOST}

    result = (info or {}).get("result") or {}
    klippy = result.get("klippy_state") or "unknown"
    out = {
        "online": True,
        "host": HOST,
        "klippy_state": klippy,
        "klippy_connected": bool(result.get("klippy_connected")),
        "error": None,
    }

    # Klipper can be in error/shutdown while Moonraker itself answers fine -
    # that distinction is the whole point of showing klippy_state separately.
    if klippy != "ready":
        out["state"] = klippy
        return out

    try:
        data = _get_json(f"{MOONRAKER}/printer/objects/query?{QUERY}")
    except Exception as e:
        out["error"] = f"{type(e).__name__}: {e}"
        out["state"] = "unknown"
        return out

    status = ((data or {}).get("result") or {}).get("status") or {}
    stats = status.get("print_stats") or {}
    display = status.get("display_status") or {}

    progress = _num(display.get("progress"))
    elapsed = _num(stats.get("print_duration"))

    # Progress-rate ETA rather than the slicer's estimate: it needs no file
    # metadata call and self-corrects as the print goes. Meaningless in the
    # first moments, so it stays None until there is something to extrapolate.
    eta = None
    if progress > 0.01 and elapsed > 30:
        eta = int(elapsed * (1 - progress) / progress)

    out.update({
        "state": stats.get("state") or "standby",
        "filename": stats.get("filename") or "",
        # print_stats.message only - it is the real state/error text. The one
        # on display_status is the LCD's M117 line, which keeps whatever the
        # last job set until something overwrites it: this machine sits idle
        # reading "Printing", and surfacing that would be a banner that lies.
        "message": stats.get("message") or "",
        "lcd_message": display.get("message") or "",
        "progress": round(progress, 4),
        "print_duration": int(elapsed),
        "total_duration": int(_num(stats.get("total_duration"))),
        "filament_used_mm": round(_num(stats.get("filament_used")), 1),
        "eta_seconds": eta,
        "extruder": _heater(status.get("extruder") or {}),
        "bed": _heater(status.get("heater_bed") or {}),
        "position": (status.get("toolhead") or {}).get("position") or [],
    })
    return out


@printer.route("/config", methods=["GET"])
@require_token
def get_config():
    """What the front end needs to decide what it can render."""
    return jsonify({
        "host": HOST,
        "camera_url_available": True,
        "ui_url": UI_URL,
        "lan_url": MOONRAKER,
        "max_streams": MAX_STREAMS,
    })


@printer.route("/status", methods=["GET"])
@require_token
def get_status():
    return jsonify(read_status())


@printer.route("/snapshot", methods=["GET"])
@require_token
def snapshot():
    """
    One JPEG, proxied. The front end polls this rather than holding a stream
    open, because a still every second or two is what a print actually needs
    and it survives a phone locking, backgrounding and waking without leaving
    a socket behind.
    """
    try:
        req = urllib.request.Request(f"{CAMERA}/?action=snapshot",
                                     headers={"User-Agent": "OpsDeck/1.0"})
        with urllib.request.urlopen(req, timeout=SNAPSHOT_TIMEOUT) as res:
            body = res.read()
            ctype = res.headers.get("Content-Type", "image/jpeg")
    except Exception as e:
        # 503 rather than 500: the app is fine, the camera is not.
        return jsonify({"error": f"camera unreachable: {type(e).__name__}"}), 503

    return Response(body, mimetype=ctype, headers={
        "Cache-Control": "no-store, no-cache, must-revalidate",
        "Pragma": "no-cache",
    })


@printer.route("/stream", methods=["GET"])
@require_token
def stream():
    """
    The live MJPEG feed, proxied.

    Opt-in from the UI, because unlike /snapshot this holds a connection open
    for as long as the tab lives. The semaphore is released in the generator's
    finally, which Flask runs when the client disconnects.
    """
    if not _stream_slots.acquire(blocking=False):
        return jsonify({"error": f"too many live streams open (max {MAX_STREAMS}); "
                                 f"close one or use the snapshot view"}), 429

    try:
        req = urllib.request.Request(f"{CAMERA}/?action=stream",
                                     headers={"User-Agent": "OpsDeck/1.0"})
        upstream = urllib.request.urlopen(req, timeout=SNAPSHOT_TIMEOUT)
    except Exception as e:
        _stream_slots.release()
        return jsonify({"error": f"camera unreachable: {type(e).__name__}"}), 503

    ctype = upstream.headers.get("Content-Type",
                                 "multipart/x-mixed-replace; boundary=boundarydonotcross")

    def pump():
        try:
            while True:
                chunk = upstream.read(8192)
                if not chunk:
                    break
                yield chunk
        except (GeneratorExit, OSError):
            pass
        finally:
            try:
                upstream.close()
            finally:
                _stream_slots.release()

    return Response(pump(), mimetype=ctype, headers={
        "Cache-Control": "no-store, no-cache, must-revalidate",
        "Pragma": "no-cache",
        # Belt and braces: an intermediary buffering an MJPEG stream turns a
        # live view into a stall.
        "X-Accel-Buffering": "no",
    })

File: static/js/printer.js
// Printer: the camera and the machine state, plus the real Fluidd UI embedded.
//
// The camera image is fetched with the API client and turned into a blob URL
// rather than pointed at ?token=... in an <img src>. Same pixels, but the API
// token stays out of the URL bar, out of history and out of any Referer -
// which matters more here than usual because this is the one section whose
// images a browser is asked to load repeatedly.
//
// Live streaming is opt-in for the same reason the server caps it: a held-open
// MJPEG socket is a real cost, and a still every couple of seconds is what
// watching a 6-hour print actually needs.

let prState = null;
let prConfig = null;
let prTimer = null;        // status + snapshot poll
let prLive = false;        // live MJPEG instead of polled stills
let prShowUI = false;      // Fluidd iframe expanded
let prBlobUrl = null;      // current snapshot object URL, revoked on replace
let prCamOk = true;

const PR_POLL_MS = 2000;

const PR_STATE_COLOR = {
  printing: "c-green", paused: "c-amber", complete: "c-blue",
  cancelled: "c-gray", error: "c-red", standby: "c-gray",
  shutdown: "c-red", disconnected: "c-red",
};

function prDuration(seconds) {
  if (seconds === null || seconds === undefined) return "—";
  const s = Math.max(0, Math.round(seconds));
  const h = Math.floor(s / 3600), m = Math.floor((s % 3600) / 60);
  if (h) return `${h}h ${String(m).padStart(2, "0")}m`;
  if (m) return `${m}m ${String(s % 60).padStart(2, "0")}s`;
  return `${s}s`;
}

function prTemp(h) {
  if (!h) return "—";
  return h.target > 0
    ? `${h.temp.toFixed(1)}° <span class="ac-dim">/ ${h.target.toFixed(0)}°</span>`
    : `${h.temp.toFixed(1)}°`;
}

// Leaving this section must stop the polling. Without it, switching to Today
// leaves a timer hitting the printer forever and quietly holding the tab's
// battery hostage on a phone.
function stopPrinterPolling() {
  if (prTimer) { clearInterval(prTimer); prTimer = null; }
  if (prBlobUrl) { URL.revokeObjectURL(prBlobUrl); prBlobUrl = null; }
}

async function renderPrinter() {
  const panel = el("panel");
  panel.innerHTML = `<div class="loading">Loading…</div>`;
  stopPrinterPolling();

  try {
    prConfig = prConfig || await API.get("/printer/config");
  } catch (e) {
    panel.innerHTML = `<p class="empty-state">Printer module unavailable.</p>`;
    return;
  }
  prState = await API.get("/printer/status");

  panel.innerHTML = `
    <div class="section-head-row">
      <h1 class="section-title">Printer</h1>
      <div class="head-actions">
        <span id="pr-state-chip"></span>
        <a class="btn" href="${escAttr(prConfig.ui_url || prConfig.lan_url)}"
           target="_blank" rel="noopener">Open Fluidd</a>
      </div>
    </div>

    <div class="pr-grid">
      <div class="joint-card pr-cam-card">
        <div class="pr-cam-head">
          <span class="block-title">Camera</span>
          <button class="chip ${prLive ? "on c-amber" : ""}" id="pr-live">
            ${prLive ? "live" : "stills"}
          </button>
          <span class="pr-cam-note" id="pr-cam-note"></span>
        </div>
        <div class="pr-cam" id="pr-cam">
          <div class="pr-cam-placeholder">connecting…</div>
        </div>
      </div>

      <div class="joint-card pr-status-card" id="pr-status"></div>
    </div>

    <div class="block-title-row">
      <h2 class="block-title">Full dashboard</h2>
    </div>
    <div class="field-row-inline">
      <button class="btn" id="pr-toggle-ui">
        ${prShowUI ? "Hide" : "Show"} the Fluidd dashboard
      </button>
      <span class="settings-hint">Embedded over the tailnet. Full control —
      everything the printer's own page can do.</span>
    </div>
    <div id="pr-ui-wrap"></div>`;

  el("pr-live").addEventListener("click", () => {
    prLive = !prLive;
    renderPrinter();
  });
  el("pr-toggle-ui").addEventListener("click", () => {
    prShowUI = !prShowUI;
    renderPrinter();
  });

  paintStatus();
  mountCamera();
  renderFluidd();

  // One timer drives both the status card and, in stills mode, the image.
  prTimer = setInterval(async () => {
    if (activeSection !== "printer") { stopPrinterPolling(); return; }
    try { prState = await API.get("/printer/status"); } catch (e) { return; }
    paintStatus();
    if (!prLive) refreshSnapshot();
  }, PR_POLL_MS);
}

function paintStatus() {
  const s = prState || {};
  const chip = el("pr-state-chip");
  const label = s.online ? (s.state || "unknown") : "offline";
  if (chip) {
    chip.className = `chip static ${s.online ? (PR_STATE_COLOR[label] || "c-gray") : "c-red"}`;
    chip.textContent = label;
  }

  const box = el("pr-status");
  if (!box) return;

  if (!s.online) {
    box.innerHTML = `
      <div class="block-title">Offline</div>
      <p class="settings-hint">No answer from ${esc(s.host || "the printer")}.
      That is the normal look when the machine is powered down — this tab will
      pick it up again on its own once it is back.</p>
      <div class="pr-err mono">${esc(s.error || "")}</div>`;
    return;
  }

  if (s.klippy_state && s.klippy_state !== "ready") {
    box.innerHTML = `
      <div class="block-title">Klipper: ${esc(s.klippy_state)}</div>
      <p class="settings-hint">Moonraker is answering but Klipper is not ready —
      usually a shutdown or a config error. Open Fluidd to see the reason and
      restart the firmware.</p>`;
    return;
  }

  const printing = s.state === "printing" || s.state === "paused";
  const pct = Math.round((s.progress || 0) * 100);

  box.innerHTML = `
    <div class="pr-temps">
      <div class="pr-temp">
        <div class="h-label">Nozzle</div>
        <div class="pr-temp-val">${prTemp(s.extruder)}</div>
      </div>
      <div class="pr-temp">
        <div class="h-label">Bed</div>
        <div class="pr-temp-val">${prTemp(s.bed)}</div>
      </div>
    </div>

    ${printing || s.filename ? `
      <div class="pr-job">
        <div class="pr-file mono" title="${escAttr(s.filename)}">${esc(s.filename || "—")}</div>
        <div class="ac-bar pr-bar"><div class="ac-bar-fill" style="width:${pct}%"></div></div>
        <div class="pr-job-meta">
          <span><strong>${pct}%</strong></span>
          <span>elapsed ${prDuration(s.print_duration)}</span>
          <span>${s.eta_seconds !== null && s.eta_seconds !== undefined
            ? "left ~" + prDuration(s.eta_seconds) : "left —"}</span>
          <span>${s.filament_used_mm ? (s.filament_used_mm / 1000).toFixed(2) + " m" : ""}</span>
        </div>
      </div>` : `
      <p class="settings-hint">Idle — nothing queued. Temperatures above are
      live.</p>`}
    ${s.message ? `<div class="pr-msg">${esc(s.message)}</div>` : ""}`;
}

// ---------------------------------------------------------------- camera

function mountCamera() {
  const box = el("pr-cam");
  const note = el("pr-cam-note");
  if (!box) return;

  if (prLive) {
    // The one place a token in the URL is unavoidable: an <img> streaming
    // multipart cannot carry a custom header. Same-origin, tailnet-only.
    const url = `/api/printer/stream?token=${encodeURIComponent(window.OPSDECK.token)}`;
    box.innerHTML = `<img id="pr-img" alt="Printer camera (live)" src="${escAttr(url)}">`;
    note.textContent = "holding one connection open";
    el("pr-img").addEventListener("error", () => {
      box.innerHTML = `<div class="pr-cam-placeholder">Live stream unavailable.</div>`;
    });
    return;
  }

  box.innerHTML = `<img id="pr-img" alt="Printer camera">
    <div class="pr-cam-placeholder" id="pr-cam-ph">connecting…</div>`;
  note.textContent = `refreshing every ${PR_POLL_MS / 1000}s`;
  refreshSnapshot();
}

async function refreshSnapshot() {
  const img = el("pr-img");
  if (!img) return;
  let blob;
  try {
    const res = await fetch("/api/printer/snapshot", {
      headers: { "X-API-Token": window.OPSDECK.token },
    });
    if (!res.ok) throw new Error(String(res.status));
    blob = await res.blob();
  } catch (e) {
    if (prCamOk) {
      prCamOk = false;
      const ph = el("pr-cam-ph");
      if (ph) ph.textContent = "Camera unreachable.";
    }
    return;
  }
  prCamOk = true;
  el("pr-cam-ph")?.remove();
  const next = URL.createObjectURL(blob);
  img.src = next;
  // Revoke the previous frame only after swapping, or the visible image is
  // torn out from under the browser mid-decode and flickers.
  if (prBlobUrl) URL.revokeObjectURL(prBlobUrl);
  prBlobUrl = next;
}

// ---------------------------------------------------------------- Fluidd

function renderFluidd() {
  const wrap = el("pr-ui-wrap");
  if (!wrap) return;
  if (!prShowUI) { wrap.innerHTML = ""; return; }

  if (!prConfig.ui_url) {
    wrap.innerHTML = `<p class="empty-state">No HTTPS front is configured for the
      printer UI, so it cannot be embedded — open it directly instead.</p>`;
    return;
  }
  wrap.innerHTML = `
    <div class="pr-ui-frame">
      <iframe src="${escAttr(prConfig.ui_url)}" title="Fluidd"
              referrerpolicy="no-referrer"></iframe>
    </div>
    <p class="settings-hint">If this stays blank, the tailnet HTTPS front for the
    printer is down — <a href="${escAttr(prConfig.ui_url)}" target="_blank"
    rel="noopener">open it in a tab</a> to check.</p>`;
}
